devito/types/dense.py

```diff
@@ -1454,3 +1453,99 @@ def parent(self):
         return self._parent
 
     _pickle_kwargs = Function._pickle_kwargs + ['parent']
+
+
+class TempFunction(DiscreteFunction):
+
+    is_TempFunction = True
+
+    def __init_finalize__(self, *args, **kwargs):
+        super().__init_finalize__(*args, **kwargs)
+
+        self._pointer_dim = kwargs.get('pointer_dim')
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        if pointer_dim not in dimensions:
+            # This is a bit hacky but it does work around duplicate dimensions when
+            # it gets to pickling
+            dimensions = as_tuple(pointer_dim) + dimensions
+        return dimensions, dimensions
+
+    def __halo_setup__(self, **kwargs):
+        pointer_dim = kwargs.get('pointer_dim')
+        dimensions = as_tuple(kwargs['dimensions'])
+        halo = as_tuple(kwargs.get('halo'))
+        if halo is None:
+            halo = tuple((0, 0) for _ in dimensions)
+        if pointer_dim is not None:
+            halo = ((0, 0),) + halo
+        return halo
+
+    @property
+    def data(self):
+        # Any attempt at allocating data by the user should fail miserably
+        raise TypeError("TempFunction cannot allocate data")
+
+    data_domain = data
+    data_with_halo = data
+    data_ro_domain = data
+    data_ro_with_halo = data
+
+    @property
+    def pointer_dim(self):
+        return self._pointer_dim
+
+    @property
+    def dim(self):
+        return self.pointer_dim
+
+    @property
+    def shape(self):
+        return self.symbolic_shape
+
+    shape_with_halo = shape
+    shpe_allocated = shape
+
+    def make(self, shape, initializer=None):
+        """
+        Create a Function which can be used to override this TempFunction
+        in a call to `op.apply(...)`.
+
+        Parameters
+        ----------
+        shape : tuple of ints
+            Shape of the domain region in grid points.
+        initializer : callable or any object exposing the buffer interface, optional
+            Data initializer. If a callable is provided, data is allocated lazily.
+        """
+        # Sanity check
+        if len(shape) != self.ndim:
+            raise ValueError("`shape` must contain %d integers, not %d"
+                             % (self.ndim, len(shape)))
+
+        return Function(name=self.name, dtype=self.dtype, dimensions=self.dimensions,
```

Comment:
What if any of the dimensions is time?

Reply:
undefined behaviour. But temporaries are never created along time, so we should be fine

Reply:
Can we add a check there is no time dim for safety?

Reply:
Done, added a few lines above in `__indices_setup__` , and I've actually made it stronger by checking for `not any(i.is_NonlinearDerived ...)`